scripts/npm/check-node-modules.js

```diff
@@ -0,0 +1,175 @@
+'use strict';
+
+// Imports
+var fs = require('fs');
+var path = require('path');
+
+// Constants
+var NPM_SHRINKWRAP_FILE = 'npm-shrinkwrap.json';
+var NPM_SHRINKWRAP_CACHED_FILE = 'node_modules/npm-shrinkwrap.cached.json';
+var FS_OPTS = {encoding: 'utf-8'};
+var PROJECT_ROOT = path.join(__dirname, '../../');
+
+// Variables
+var progressIndicatorTimer = null;
+
+// Run
+_main();
+
+// Functions - Definitions
+function _main() {
+  var reinstallIfStale = process.argv.indexOf('--reinstall') !== -1;
+  checkNodeModules(reinstallIfStale);
+}
+
+function cacheNpmShrinkwrapFile() {
+  var absoluteMarkerFilePath = path.join(PROJECT_ROOT, NPM_SHRINKWRAP_FILE);
+  var absoluteCachedMarkerFilePath = path.join(PROJECT_ROOT, NPM_SHRINKWRAP_CACHED_FILE);
+
+  startProgressIndicator('    Caching marker file...');
+  copyFile(absoluteMarkerFilePath, absoluteCachedMarkerFilePath, onCopied);
+
+  // Helpers
+  function onCopied(err) {
+    stopProgressIndicator();
+    if (err) logError(err);
+  }
+}
+
+function checkNodeModules(reinstallIfStale) {
+  var nodeModulesOk = compareMarkerFiles();
+
+  if (nodeModulesOk) {
+    console.log(':-) npm dependencies are looking good!');
+  } else {
+    console.warn(':-( npm dependencies are stale or in an unknown state!');
+
+    if (reinstallIfStale) {
+      purgeModules();
+      installModules();
+    }
+  }
+
+  return nodeModulesOk;
+}
+
+function compareMarkerFiles() {
+  var absoluteMarkerFilePath = path.join(PROJECT_ROOT, NPM_SHRINKWRAP_FILE);
+  var absoluteCachedMarkerFilePath = path.join(PROJECT_ROOT, NPM_SHRINKWRAP_CACHED_FILE);
+
+  if (!fs.existsSync(absoluteCachedMarkerFilePath)) return false;
+
+  var markerContent = fs.readFileSync(absoluteMarkerFilePath, FS_OPTS);
+  var cachedMarkerContent = fs.readFileSync(absoluteCachedMarkerFilePath, FS_OPTS);
+
+  return markerContent === cachedMarkerContent;
+}
+
+// Implementation based on
+// https://stackoverflow.com/questions/11293857/fastest-way-to-copy-file-in-node-js#answer-21995878
+function copyFile(srcPath, dstPath, callback) {
+  var callbackCalled = false;
+
+  if (!fs.existsSync(srcPath)) {
+    done(new Error('Missing source file: ' + srcPath));
+    return;
+  }
+
+  var rs = fs.createReadStream(srcPath);
+  rs.on('error', done);
+
+  var ws = fs.createWriteStream(dstPath);
+  ws.on('error', done);
+  ws.on('finish', done);
+
+  rs.pipe(ws);
+
+  // Helpers
+  function done(err) {
+    if (callback && !callbackCalled) {
+      callbackCalled = true;
+      callback(err);
+    }
+  }
+}
+
+// Custom implementation of `rm -rf` that works consistently across OSes
+function deleteDirSync(path) {
+  if (fs.existsSync(path)) {
+    fs.readdirSync(path).forEach(deleteDirOrFileSync);
+    fs.rmdirSync(path);
+  }
+
+  // Helpers
+  function deleteDirOrFileSync(subpath) {
+    var curPath = path + '/' + subpath;
+
+    if (fs.lstatSync(curPath).isDirectory()) {
+      deleteDirSync(curPath);
+    } else {
+      fs.unlinkSync(curPath);
+    }
+  }
+}
```

Comment:
Would we care to use `rm -rf` on *NIX based systems for performance reasons ?
I didn't bother, because this will be running only rarely and then again `npm install` is the main bottleneck.